File: src/Core/StagingFinancialEvents/Services/Transformers/EventProcessors/StockEventProcessor.py
from typing import Sequence

import Core.FinancialEvents.Schemas.Grouping as pgf
from Core.StagingFinancialEvents.Contracts.EventProcessor import EventProcessor
from Core.StagingFinancialEvents.Schemas.Events import (
    StagingTradeEvent,
    StagingTradeEventStockAcquired,
)


class StockEventProcessor(EventProcessor[StagingTradeEvent, pgf.TradeEventStockAcquired | pgf.TradeEventStockSold]):

    # TODO: Create trade events based on corporate events
    def createMissingStockTradesFromCorporateActions(
        self,
    ) -> Sequence[pgf.TradeEventStockAcquired | pgf.TradeEventStockSold]:
        return []

    def process(self, input: StagingTradeEvent) -> pgf.TradeEventStockAcquired | pgf.TradeEventStockSold:
        if isinstance(input, StagingTradeEventStockAcquired):
            converted = pgf.TradeEventStockAcquired(
                ID=input.ID,
                ISIN=input.ISIN,
                Ticker=input.Ticker or "",
                AssetClass=input.AssetClass,
                Date=input.Date,
                Multiplier=input.Multiplier,
                AcquiredReason=input.AcquiredReason,
                ExchangedMoney=input.ExchangedMoney,
            )
            return converted

        converted = pgf.TradeEventStockSold(
            ID=input.ID,
            ISIN=input.ISIN,
            Ticker=input.Ticker or "",
            AssetClass=input.AssetClass,
            Date=input.Date,
            Multiplier=input.Multiplier,
            ExchangedMoney=input.ExchangedMoney,
        )
        return converted

File: lib/core/stagingProcessor/transformers/StockEventProcessor.ts
import { FinancialIdentifier } from "@brrr/core/schemas/FinancialIdentifier.ts";
import type { TradeEventStockAcquired, TradeEventStockSold } from "@brrr/core/schemas/Events.ts";
import type { StagingTradeEventStock } from "@brrr/core/schemas/staging/Events.ts";

export function processStockEvent(
	input: StagingTradeEventStock,
): TradeEventStockAcquired | TradeEventStockSold {
	const identifier = FinancialIdentifier.fromStagingIdentifier(input.financialIdentifier);
	if (input.kind === "StagingStockAcquired") {
		return {
			kind: "StockAcquired",
			id: input.id,
			financialIdentifier: identifier,
			assetClass: input.assetClass,
			date: input.date,
			multiplier: input.multiplier,
			exchangedMoney: input.exchangedMoney,
			acquiredReason: input.acquiredReason,
			provenance: [],
		};
	}
	return {
		kind: "StockSold",
		id: input.id,
		financialIdentifier: identifier,
		assetClass: input.assetClass,
		date: input.date,
		multiplier: input.multiplier,
		exchangedMoney: input.exchangedMoney,
		provenance: [],
	};
}
